planetiler-core/src/main/java/com/onthegomap/planetiler/geo/TileExtents.java

```diff
@@ -52,9 +59,9 @@ public boolean test(TileCoord tileCoord) {
 
   /**
    * X/Y extents within a given zoom level. {@code minX} and {@code minY} are inclusive and {@code maxX} and {@code
-   * maxY} are exclusive.
+   * maxY} are exclusive. shape is an optional polygon defining a more refine shape
    */
-  public record ForZoom(int minX, int minY, int maxX, int maxY) {
+  public record ForZoom(int minX, int minY, int maxX, int maxY, MultiPolygon shape) {
```

Comment:
One contributing factor to making pass2 slow is that we're projecting tile coordinates back to lat/lon, and also doing an expensive "intersects" test on a raw JTS geometry.  In order to speed this up:

1) we could project the shape to world web mercator coordinates (and possibly create a separate instance scaled per zoom level to go into `ForZoom`)
2) we could construct a `PreparedGeometry` ahead of time that's optimized for queries like `intersects`

Reply:
You should be able to use `GeoUtils.latLonToWorldCoords`